### PR TITLE
Added support for manga list and anime/manga list entry update

### PR DESCRIPTION
**Pull request by naps250**
Added support for user's manga list.
Added support for updating user's anime and manga MAL records.
Minor renaming and reordering.
Added 'Anime' and 'Manga' folders (and namespaces) for better organization of the 'MalApi' project.

### CODE REVIEW

**Comment by LHCGreg** [comment]
Thanks for your contribution. I've glanced through the changes. I'll take a deeper look this weekend and will have more comments then but I have some comments now.

First, the creation of the MalApi.Anime and MalApi.Manga namespaces. I can see how one might like how that organizes the MalApi code, especially with several parallel anime and manga types, but I don't know if that's something a library client would want. Now you pretty much have to import two namespaces to do anything. There is not such a huge number of types in this library and moving some types into another namespace is a backwards compatibility breaking change. Some breaking changes I think will be necessary like the renaming of CompletionStatus -> AnimeCompletionStatus due to my lack of foresight, but it should be kept to a minimum.

Secondly, tests! The new functionality should have some coverage in unit/integration tests. I realize integration tests for functionality that makes updates to a user's list poses some problems regarding credentials, but it can be done, for example by reading the username and password from environment variables and documenting what kind of setup is required to run the tests as far as setting the environment variables and setting up the MAL account used in the tests.

**Comment by naps250** [comment]
Hi,

Thank you for you feedback. I really appreciate it.

I reverted the namespace change but still kept the new folders. I had completely overlooked the backward compatibility point of view when initially making this change.
I also went through the .cs files and removed any unused using statements.

Both of the above are part of the second commit.

I will start working on the test now while waiting for your further comments.

**Comment by LHCGreg** [comment]
Please revert the removal of unused using directives. The changes clutter the PR and are unrelated to manga and update functionality. I find it useful to have common namespaces like System.Collections.Generic and System.Linq already imported.

File: MalApi/Anime/UpdateAnimeObject.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
+﻿using System.Xml.Serialization;
+
+namespace MalApi
+{
+    [XmlRoot("entry")]
+    public class UpdateAnimeObject : UpdateObjectBase
```
Comment (LHCGreg) [inline MalApi/Anime/UpdateAnimeObject.cs]: "Object" is kind of generic. "AnimeUpdate" would describe the purpose as well.

Comment (LHCGreg) [reply MalApi/Anime/UpdateAnimeObject.cs]: The properties in this class are all strings. The users of the library should be dealing with integers for episode count, AnimeCompletionStatus for status, etc, instead of expecting them to serialize those properties to strings.

File: MalApi/Anime/UpdateAnimeObject.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+                                 string timesRewatched = null,
+                                 string rewatchValue = null,
+                                 string dateStart = null,
+                                 string dateFinish = null,
+                                 string priority = null,
+                                 string enableDiscussion = null,
+                                 string enableRewatching = null,
+                                 string comments = null,
+                                 string tags = null)
+        {
```
Comment (LHCGreg) [inline MalApi/Anime/UpdateAnimeObject.cs]: Is this constructor adding any value over calling the default constructor and then setting properties?

File: MalApi/Anime/UpdateAnimeObject.cs (hunk 1 above)
Comment (LHCGreg) [reply MalApi/Anime/UpdateAnimeObject.cs]: What does it mean if a property is not set? Does it not get updated or does it get set to an empty value in MAL? This should be documented in a doc comment on the class.

Comment (LHCGreg) [reply MalApi/Anime/UpdateAnimeObject.cs]: Might be nice to link to the MAL documentation at https://myanimelist.net/modules.php?go=api#animevalues in a comment.

File: MalApi/MyAnimeListApi.cs [hunk 3]
```diff
@@ -227,6 +305,147 @@ public Task<MalUserLookupResults> GetAnimeListForUserAsync(string user)
...
         /// <summary>
+        /// Updates a user's anime list entry.
+        /// </summary>
+        /// <param name="animeId">ID of the anime</param>
+        /// <param name="updateInfo">The updated information</param>
+        /// <param name="cancellationToken"></param>
+        /// <param name="user">MAL user</param>
+        /// <param name="password">MAL password</param>
+        /// <returns></returns>
+        public async Task<string> UpdateAnimeForUserAsync(int animeId, UpdateAnimeObject updateInfo, string user, string password, CancellationToken cancellationToken)
```
Comment (LHCGreg) [inline MalApi/MyAnimeListApi.cs]: The return type should be `Task` instead of `Task<string>` with an always null string.

File: MalApi/MyAnimeListApi.cs [hunk 4]
```diff
@@ -16,7 +15,14 @@ namespace MalApi
     /// </summary>
     public class MyAnimeListApi : IMyAnimeListApi
     {
-        private const string MalAppInfoUri = "https://myanimelist.net/malappinfo.php?status=all&type=anime";
+        private static readonly string AnimeDetailsUrlFormat = "https://myanimelist.net/anime/{0}";
+
+        private const string MalAppAnimeInfoUri = "https://myanimelist.net/malappinfo.php?status=all&type=anime&u={0}";
+        private const string MalAppMangaInfoUri = "https://myanimelist.net/malappinfo.php?status=all&type=manga&u={0}";
+
+        private const string MalAppAnimeUpdateUri = "https://myanimelist.net/api/animelist/update/{0}.xml";
```
Comment (LHCGreg) [inline MalApi/MyAnimeListApi.cs]: No need for "App" in the name since it has nothing to do with malappinfo.php. Also, this is a format string rather than a string that can be used directly, but the name does not indicate this. Consider moving the format string directly into the function that uses it so that there is no distance between the format string and the string.Format call. That way it is clear where the format arguments get put and how many are needed.

File: MalApi/MyAnimeListApi.cs [hunk 5]
```diff
@@ -227,6 +305,147 @@ public Task<MalUserLookupResults> GetAnimeListForUserAsync(string user)
...
+        /// </summary>
+        /// <param name="animeId">ID of the anime</param>
+        /// <param name="updateInfo">The updated information</param>
+        /// <param name="cancellationToken"></param>
+        /// <param name="user">MAL user</param>
+        /// <param name="password">MAL password</param>
+        /// <returns></returns>
+        public async Task<string> UpdateAnimeForUserAsync(int animeId, UpdateAnimeObject updateInfo, string user, string password, CancellationToken cancellationToken)
+        {
+            string userInfoUri = string.Format(MalAppAnimeUpdateUri, Uri.EscapeDataString(animeId.ToString()));
```
Comment (LHCGreg) [inline MalApi/MyAnimeListApi.cs]: Uri.EscapeDataString is not really necessary, but ok.

File: MalApi/MyAnimeListApi.cs [hunk 6]
```diff
@@ -227,6 +305,147 @@ public Task<MalUserLookupResults> GetAnimeListForUserAsync(string user)
...
+                using (TextReader xmlTextReader = new StringReader(response))
+                {
+                    try
+                    {
+                        return xmlTextReader.ToString();
+                    }
+                    catch (Exception ex)
+                    {
+                        throw new Exception(string.Format("Exception for user {0} and anime ID {1}.", user, animeId), ex);
+                    }
```
Comment (LHCGreg) [inline MalApi/MyAnimeListApi.cs]: responseProcessingFunc should not be catching exceptions and rethrowing them as a general Exception. The baseErrorMessage parameter to ProcessRequestAsync is used as part of the error handling in ProcessRequestAsync.

Comment (LHCGreg) [reply MalApi/MyAnimeListApi.cs]: https://myanimelist.net/modules.php?go=api says the response is 'Updated' or detailed error message. Have you tested this? If there is an error does it return an HTTP error status code or does it return an OK HTTP status code? If there is an error message, is it a plain string or is it in something like xml or json? It looks like the code here does nothing to surface the MAL error message to the caller.

File: MalApi/MyAnimeListApi.cs [hunk 7]
```diff
@@ -227,6 +305,147 @@ public Task<MalUserLookupResults> GetAnimeListForUserAsync(string user)
...
+
+            Logging.Log.InfoFormat("Updating anime entry for MAL anime ID {0}, user {1} using URI {2}", animeId, user, userInfoUri);
+
+            Func<string, string> responseProcessingFunc = (response) =>
+            {
+                using (TextReader xmlTextReader = new StringReader(response))
+                {
+                    try
+                    {
+                        return xmlTextReader.ToString();
```
Comment (LHCGreg) [inline MalApi/MyAnimeListApi.cs]: If responseProcessingFunc has nothing useful to return, which makes sense for an update API call, responseProcessingFunc should have a return type of object and return null.

File: MalApi/MyAnimeListApi.cs [hunk 8]
```diff
@@ -72,6 +78,40 @@ private HttpRequestMessage InitNewRequest(string uri, HttpMethod method)
             return request;
         }
 
+        private HttpRequestMessage InitNewRequestWithCredentials(string user, string password, string uri, HttpMethod method, UpdateObjectBase updateInfo = null)
```
Comment (LHCGreg) [inline MalApi/MyAnimeListApi.cs]: Consider moving uri and method to be the first two parameters, as they are kind of the "primary" parameters of the method.

File: MalApi/MyAnimeListApi.cs [hunk 9]
```diff
@@ -72,6 +78,40 @@ private HttpRequestMessage InitNewRequest(string uri, HttpMethod method)
...
+                credentials =  Convert.ToBase64String(plainTextBytes);
+            }
+            else
+            {
+                throw new Exception("Invalid input credentials format (both username and password have to be input OR just the base64 representation");
+            }
+
+            if (UserAgent != null)
+            {
+                request.Headers.TryAddWithoutValidation("User-Agent", UserAgent);
```
Comment (LHCGreg) [inline MalApi/MyAnimeListApi.cs]: You can avoid repeating the logic that is in InitNewRequest by calling InitNewRequest and then adding onto the request it returns.

File: MalApi/MyAnimeListApi.cs (hunk 8 above)
Comment (LHCGreg) [reply MalApi/MyAnimeListApi.cs]: updateInfo does not belong in InitNewRequestWithCredentials. When functionality for adding and removing of anime/manga entries is added, that will use credentials but will not take the same input that updating does. Any MAL endpoint-specific logic can be done in the method for that endpoint on top of the HttpRequestMessage object returned by InitNewRequest or InitNewRequestWithCredentials.

File: MalApi/MyAnimeListApi.cs [hunk 10]
```diff
@@ -72,6 +78,40 @@ private HttpRequestMessage InitNewRequest(string uri, HttpMethod method)
...
+                // Adding the authorization header with the credentials
+                request.Headers.TryAddWithoutValidation("Authorization", "Basic " + credentials);
```
Comment (LHCGreg) [inline MalApi/MyAnimeListApi.cs]: Why only set the Authorization header if the UserAgent property is set?

Comment (LHCGreg) [reply MalApi/MyAnimeListApi.cs]: Using header properties that the HttpRequestMessage API supports directly is preferable to using TryAddWithoutValidation. TryAddWithoutValidation is used for User-Agent because it complains if the user agent you set does not follow the same format browsers use, which is not the case when it is being (mis)appropriated as an API key. For authorization you should be able to use

```
request.Headers.Authorization = new AuthenticationHeaderValue("Basic", credentials);
```

File: MalApi/MyAnimeListApi.cs [hunk 11]
```diff
@@ -72,6 +78,40 @@ private HttpRequestMessage InitNewRequest(string uri, HttpMethod method)
...
+        private HttpRequestMessage InitNewRequestWithCredentials(string user, string password, string uri, HttpMethod method, UpdateObjectBase updateInfo = null)
+        {
+            HttpRequestMessage request = new HttpRequestMessage(method, uri);
+
+            // Requests with credentials require them to be encoded in base64
+            string credentials;
+            if (user != null && password != null)
+            {
+                var plainTextBytes = Encoding.UTF8.GetBytes(user + ":" + password);
+                credentials =  Convert.ToBase64String(plainTextBytes);
```
Comment (LHCGreg) [inline MalApi/MyAnimeListApi.cs]: There's an extra space after the equals sign.

File: MalApi/MyAnimeListApi.cs [hunk 12]
```diff
@@ -72,6 +78,40 @@ private HttpRequestMessage InitNewRequest(string uri, HttpMethod method)
...
+            }
+            else
+            {
+                throw new Exception("Invalid input credentials format (both username and password have to be input OR just the base64 representation");
```
Comment (LHCGreg) [inline MalApi/MyAnimeListApi.cs]: Instead of Exception, ArgumentNullException would be more appropriate. In order to call the ArgumentNullException constructor that takes a parameter name, you'll have to check user and password one at a time.

File: MalApi/MyAnimeListApi.cs (hunk 3 above)
Comment (LHCGreg) [reply MalApi/MyAnimeListApi.cs]: Please change "user" to "username" to be more clear.

File: MalApi/MyAnimeListApi.cs (hunk 12 above)
Comment (LHCGreg) [reply MalApi/MyAnimeListApi.cs]: Also, I am confused by the "OR just the base64 representation" in the error message.

File: MalApi/MyAnimeListApi.cs [hunk 13]
```diff
@@ -72,6 +78,40 @@ private HttpRequestMessage InitNewRequest(string uri, HttpMethod method)
...
+                // Adding the authorization header with the credentials
+                request.Headers.TryAddWithoutValidation("Authorization", "Basic " + credentials);
+            }
+
+            // Encoding and adding the new information in the content(body) of the request
+			if(updateInfo != null)
+			{
+				var xml = updateInfo.GenerateXml();
+				var content = "data=" + WebUtility.UrlEncode(xml);
+				request.Content = new StringContent(content, Encoding.UTF8, "application/x-www-form-urlencoded");
```
Comment (LHCGreg) [inline MalApi/MyAnimeListApi.cs]: Does this not work? The FormUrlEncodedContent class exists so you don't have to encode it yourself.

```
request.Content = new FormUrlEncodedContent(new KeyValuePair<string, string>[] { new KeyValuePair<string, string>("data", xml) });
```

**Comment by LHCGreg** [comment]
Added comments for anime update functionality. Most or all of it applies to manga update functionality as well.

File: MalApi/Anime/UpdateAnimeObject.cs (hunk 1 above)
Comment (naps250) [reply MalApi/Anime/UpdateAnimeObject.cs]: I agree with all your comments.

File: MalApi/Anime/UpdateAnimeObject.cs (hunk 2 above)
Comment (naps250) [reply MalApi/Anime/UpdateAnimeObject.cs]: Not really... I'll remove it.

File: MalApi/MyAnimeListApi.cs (hunk 8 above)
Comment (naps250) [reply MalApi/MyAnimeListApi.cs]: Technically, delete takes only an id as input while add and update take the same input as far as I understand from the API. But again I agree that setting the content inside InitNewRequestWithCredentials is not the best solution due to possible future changes to the API and the functionality that the method name reflects.

File: MalApi/MyAnimeListApi.cs (hunk 12 above)
Comment (naps250) [reply MalApi/MyAnimeListApi.cs]: The error message part was a leftover from earlier ideas.

**Comment by naps250** [comment]
Thanks again for your feedback.

I addressed all of your comments regarding the functionality itself but couldn't quite get the tests set up properly. My apologies!

File: MalApi/MalAppInfoXml.cs [hunk 14]
```diff
@@ -194,6 +202,100 @@ public static MalUserLookupResults Parse(XDocument doc)
...
+            }
+
+            if (!doc.Root.HasElements)
+            {
+                throw new MalUserNotFoundException("No MAL list exists for this user.");
+            }
+
+            XElement myinfo = GetExpectedElement(doc.Root, "myinfo");
+            int userId = GetElementValueInt(myinfo, "user_id");
+            string canonicalUserName = GetElementValueString(myinfo, "user_name");
```
Comment (LHCGreg) [inline MalApi/MalAppInfoXml.cs]: This first section of the method is duplicated from ParseAnimeResults. A way to avoid the duplication would be to add another constructor to MalUserLookupResults that takes both an anime list and a manga list and have one function in MalAppInfoXml that looks for both "anime" XML elements and "manga" XML elements.

MalUserLookupResults should then have an empty list for anime or manga instead of null.

File: MalApi/Manga/MalMangaType.cs [hunk 15]
```diff
@@ -0,0 +1,14 @@
+﻿namespace MalApi
+{
+    public enum MalMangaType
+    {
+        Unknown = 0,
+        Manga = 1,
+        Novel = 2,
+        OneShot = 3,
+        Doujin = 4, // not sure
```
Comment (LHCGreg) [inline MalApi/Manga/MalMangaType.cs]: Why aren't you sure about Doujin and OEL?

File: MalApi/Anime/UpdateAnime.cs [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+
+namespace MalApi
+{
+    /// <summary>
+    /// The update object sent to MAL when updating an anime entry.
+    /// Only specified values will be changed. The rest will remain unchanged.
+    /// More details: https://myanimelist.net/modules.php?go=api#animevalues
+    /// </summary>
+    [XmlRoot("entry")]
+    public class UpdateAnime : UpdateObjectBase, IXmlSerializable
```
Comment (LHCGreg) [inline MalApi/Anime/UpdateAnime.cs]: Class names should be nouns, so this and UpdateManga should be AnimeUpdate/MangaUpdate.

Comment (LHCGreg) [reply MalApi/Anime/UpdateAnime.cs]: I thought I had commented on this yesterday, I guess I missed it. Update functionality should use System.Xml.Linq for XML generation instead of System.Xml.Serialization:

- Consistency. The MalAppInfoXml class already uses System.Xml.Linq.
- It avoids adding a dependency on System.Xml.XmlSerializer to the nuget package.
- It avoids putting implementation details in the API surface of MalApi such the XML-related attributes and the UpdateObjectBase class.

The method for creating XML should be internal, not public, and would look something like this using System.Xml.Linq.

```
            XDocument document = new XDocument(
                new XDeclaration("1.0", "UTF-8", null),
                new XElement("entry",
                    new XElement("Episode", Episode),
                    new XElement("Status", (int?)Status),
                    new XElement("Score", Score),
                   ...
                )
            );

            using (Utf8StringWriter writer = new Utf8StringWriter())
            {
                document.Save(writer);
                return writer.ToString();
            }
```

**Comment by LHCGreg** [comment]
I got a chance to look at the tests. Perhaps I was not clear when I mentioned environment variables. The idea is that username and password get stored in environment variables, *outside of the test code* so there is no risk of committing MAL credentials to git. If someone wants to run the tests, they are responsible for setting the environment variables (for example, using the [Windows environment variables editor](https://superuser.com/a/985947) outside of Visual Studio or [a .runsettings file](https://docs.microsoft.com/en-us/visualstudio/test/configure-unit-tests-by-using-a-dot-runsettings-file)). There would be an environment variable for username and password of a test account and the readme.txt would document what environment variables are required and what is expected of that test account in order for the tests to pass. For example, must have at least one anime and manga rating, must have an entry for Cowboy Bebop, whatever. So there do not need to be variables for an anime id, episode number, test score, or test status. Those are hardcoded in the tests and documented in the readme. Ideally the state of the test account when tests are not being run does not have a score on any anime or manga to avoid having any influence, no matter how small, on the overall average score.

It's not necessary to use the environment variable in the test for getting manga list if the test is just checking that the list returned is not empty. You can hardcode any user that has a non-empty manga list, like I did for the anime test. Alternatively the test could continue to use the environment variable test user and do some more thorough checks than if the list is non-empty.

The integration tests for updating should verify that the update actually happened by getting the user's list and checking that it is as expected.

The integration tests should pass no matter how many times they are run. That is, the act of running the tests should not put the test user's list into a state that causes the tests to fail the next time they are run.

The integration tests should also be meaningful tests no matter how many times they are run. If I run the update test and it only does one update, then if I run the test again, it's not really testing anything because the state of the user's list is still the same. A way to ensure that the update functionality is being tested is to do two updates and get the user's list after each.

The tests should cover as many scenarios as possible. If there is a bug with updating any particular property, they should catch that. They should catch a bug that happens when only some properties are specified in the update. They should cover not just the happy path but test that errors are handled correctly. For example, what happens if you update with a wrong password? Does the code correctly throw an exception of the correct type that indicates what was wrong? Have to be careful with this though, as MAL will IP block you for a bit if you give incorrect credentials too many times in a short period. How about if you give an invalid value for one of the properties?

There should be unit tests for the MalAppInfoXml parsing of manga lists, even if the integration tests are modified to do more than test that a returned list is non-empty, so that functionality can be tested without needing to set up a test MAL user or in case the integration tests are not able to be run for some reason. At the moment there is no test coverage for manga list XML parsing.

File: MalApi/Manga/MalMangaType.cs (hunk 15 above)
Comment (smovsesyan-rms) [reply MalApi/Manga/MalMangaType.cs]: Because I couldn't find such entries when I added that. I found a doujin and confirmed that it returns 4, leaving OEL wtih 7, so this is confirmed OK now.

Comment (naps250) [reply MalApi/Manga/MalMangaType.cs]: I didn't bother finding a MAL entry to verify it at the time of writing. I went back and found a doujin and confirmed that it returns 4. That leaves OEL as for 7, so this is OK.

**Comment by naps250** [comment]
I addressed your latest comments and will focus on the tests now.

**Comment by naps250** [comment]
In order to verify the updated information I will need to harvest some of it from the page's HTML since the API does not return everything. Do you prefer I continue to use regex like you've done for the genres or should I include a 3rd party library like the [Html Agility Pack](http://html-agility-pack.net/?z=codeplex)?

**Comment by LHCGreg** [comment]
HtmlAgilityPack is fine. Since this is just used for testing (I assume), adding a dependency is no big deal. I try to avoid adding dependencies to MalApi itself when possible but if this code later becomes part of the MalApi's public API I think I'd be ok with HtmlAgilityPack as a dependency since it's relatively small and fits the purpose of MalApi well.

**Comment by naps250** [comment]
Your assumption is correct. This is needed purely for testing. I will go ahead and use HAP.

**Comment by LHCGreg** [comment]
I thought the getting of anime/manga details for a user was going to be for testing only? Not only is it part of the MyAnimeListApi class, but it is public. Let's keep the changes to the public API in this pull request to updating only. Functionality for getting details for a particular anime/manga for a user has its own challenges:

* The MyAnimeListApi class is documented as having thread-safe methods, but logging in using cookies on an HttpClient that is a class member variable makes that no longer true.
* Logging in every time when getting anime/manga's details for a user is suboptimal.

In order to avoid adding a dependency on HtmlAgilityPack for functionality that won't be used by library users, the code for getting anime/manga's details for a user should be pulled out of the MyAnimeListApi class and into the integration test project. That code can be a little more straightforward if it doesn't need to use MyAnimeListApi's internal abstractions that do logging and exception handling.


Some other comments:

You can use

`Assert.Throws<MalApiRequestException>(() => api.UpdateAnimeForUser(animeId, partialBaseInfo, username, password));`

For asserting that some code throws a particular exception.

What happens if a user tries to update an anime id that doesn't exist or that isn't on their list? I didn't see any test coverage for that.
